datasets/code_search_net/code_search_net.py

```diff
@@ -0,0 +1,216 @@
+        ),
+        datasets.BuilderConfig(
+            name="python",
+            version=VERSION,
+            description="Pyhton language",
+        ),
+        datasets.BuilderConfig(
+            name="javascript",
+            version=VERSION,
+            description="Javascript language",
+        ),
+        datasets.BuilderConfig(
+            name="ruby",
+            version=VERSION,
+            description="Ruby language",
+        ),
+        datasets.BuilderConfig(
+            name="php",
+            version=VERSION,
+            description="PHP language",
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = "all"
+
+    def _info(self):
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=datasets.Features(
+                {
+                    "repository_name": datasets.Value("string"),
+                    "func_path_in_repository": datasets.Value("string"),
+                    "func_name": datasets.Value("string"),
+                    "whole_func_string": datasets.Value("string"),
+                    "language": datasets.Value("string"),
+                    "func_code_string": datasets.Value("string"),
+                    "func_code_tokens": datasets.Sequence(datasets.Value("string")),
+                    "func_documentation_string": datasets.Value("string"),
+                    "func_documentation_tokens": datasets.Sequence(datasets.Value("string")),
+                    "split_name": datasets.Value("string"),
+                    "func_code_url": datasets.Value("string"),
+                    # TODO - add licensing info in the examples
+                }
+            ),
+            # No default supervised keys
+            supervised_keys=None,
+            homepage=_HOMEPAGE,
+            license=_LICENSE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        """Returns SplitGenerators.
+
+        Note: The original data is stored in S3, and follows this unusual directory structure:
+            ```
+            .
+            ├── <language_name>  # e.g. python
+            │   └── final
+            │       └── jsonl
+            │           ├── test
+            │           │   └── <language_name>_test_0.jsonl.gz
+            │           ├── train
+            │           │   ├── <language_name>_train_0.jsonl.gz
+            │           │   ├── <language_name>_train_1.jsonl.gz
+            │           │   ├── ...
+            │           │   └── <language_name>_train_n.jsonl.gz
+            │           └── valid
+            │               └── <language_name>_valid_0.jsonl.gz
+            ├── <language_name>_dedupe_definitions_v2.pkl
+            └── <language_name>_licenses.pkl
+            ```
+        """
+        data_urls = _URLs[self.config.name]
+        # Extracts the language archives
+        lang_names_and_dirs = [(language, dl_manager.download_and_extract(url)) for language, url in data_urls]
```

Comment:
dl_manager.download_and_extract can take a list or a dict as input, and it will return a list or a dict accordingly.

This is useful to group the urls in lists or dicts to allow parallel downloads